### PR TITLE
IO Transit: Collection of improvements to SharedAssetInput

### PR DESCRIPTION
**Pull request by Shadowfiend**
Extracted these from the swap work and rebased against main for
holistic review.

Big ticket global change:
- `SharedSlideUpMenu` *only renders its children when open*! This has
  high potential for unexpected bugs, though I haven't found any yet,
  as I believe we weren't *depending* on the behavior of children
  rendering in spite of being invisible. The motivation here was ensuring
  React lifecycle hooks fired at a useful time for slide-up menu contents
  (see the search input auto-focus change mentioned below).

Big ticket changes for `SharedAssetInput`:
- @henryboldi's work from #847, including search by symbol for assets.
- An expansion of this work to search/filter by contract addresses
  where applicable.
- Auto-focusing of the search input when the `SharedAssetInput` asset
  selector is displayed.
- Disabled state for the whole `SharedAssetInput` (in addition to the
  existing ability to only disable asset selection while still allowing
  editing the asset amount).

A couple of small quality-of-life improvements:
- The `SharedAssetInput` type is now parametrized so that more specific
  asset lists preserve their types in the `onAssetChange` callback.
- Disabled states use HTML-native disabled states rather than working
  with pointer events.
- When available, the `key` for asset entries in the asset selector
  includes the asset's contract address. This specifically addresses
  bugs with displaying certain assets like aUSDC that have two versions
  with identical symbols (for Aave aToken v1 vs v2) but different
  contract addresses. Without this change, asset filtering would fail
  to remove these assets due to React's reliance on `key` to properly
  update lists on change, making the whole experience feel broken.

### CODE REVIEW

File: ui/components/Shared/SharedAssetInput.tsx [hunk 1]
```diff
@@ -223,16 +264,18 @@ export default function SharedAssetInput(
           </>
         ) : (
           <>
-            <div className={classNames({ disable_click: disableDropdown })}>
+            <div>
               {selectedAsset?.symbol ? (
                 <SelectedAssetButton
+                  isDisabled={isDisabled || disableDropdown}
```
Comment (henryboldi) [inline ui/components/Shared/SharedAssetInput.tsx:270]: I'm admittedly a bit confused here w/ `disableDropdown` vs `isDisabled`

File: ui/components/Shared/SharedAssetInput.tsx [hunk 2]
```diff
@@ -321,6 +365,10 @@ export default function SharedAssetInput(
           input[type="number"] {
             -moz-appearance: textfield;
           }
+          .input_amount:disabled {
```
Comment (henryboldi) [inline ui/components/Shared/SharedAssetInput.tsx:368]: 🤌👏

**Review by henryboldi — APPROVED** [review]
Looks good! 👏🧙

File: ui/components/Shared/SharedAssetInput.tsx [hunk 3]
```diff
@@ -223,16 +264,18 @@ export default function SharedAssetInput(
...
                 <SelectedAssetButton
+                  isDisabled={isDisabled || disableDropdown}
                   asset={selectedAsset}
                   toggleIsAssetMenuOpen={toggleIsAssetMenuOpen}
                 />
               ) : (
                 <SharedButton
                   type="secondary"
                   size="medium"
+                  isDisabled={isDisabled || disableDropdown}
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply ui/components/Shared/SharedAssetInput.tsx:278]: `disableDropdown` was implemented for the send flow, to fix the asset if you hit “Send” for a particular asset on the asset screen and not allow it to be changed---it still allows the amount to be edited. I'm also using it in Swap if you enter swap from the asset list.

`isDisabled` disables the whole input, including the amount.

I don't love the distinction, and I'm not 100% convinced we want both—but I was also trying not to keep going down the rabbit hole and changing behavior elsewhere in the extension 😅

File: ui/components/Shared/SharedSlideUpMenu.tsx [hunk 4]
```diff
@@ -42,7 +42,7 @@ export default function SharedSlideUpMenu(props: Props): ReactElement {
         onClick={close}
         aria-label="Close menu"
       />
-      {children}
+      {isOpen ? children : <></>}
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply ui/components/Shared/SharedSlideUpMenu.tsx:45]: Mmm, great point, probably missed it as I was moving through things quickly. Any thoughts on a good way we could fix this? Maybe something like:

```typescript
const slideUpTransitionMs = 445

...

  const [doRenderContent, setDoRenderContent] = useState(false)
  const [renderContentTimer, setRenderContentTimer] = useState(0)
  useEffect(() => {
    if (isOpen) {
      if (renderContentTimer) {
        clearTimeout(renderContentTimer)
        setRenderContentTimer(0)
      }
      setDoRenderContent(true)
    } else {
      setRenderContentTimer(setTimeout(() => setDoRenderContent(true), slideUpTransitionMs))
    }
  }, [isOpen, setDoRenderContent, renderContentTimer, setRenderContentTimer])

  ...

  {doRenderContent ? children : <></>}

  ...

            transition: transform cubic-bezier(0.19, 1, 0.22, 1) ${slideUpTransitionMs}ms;
```

?

[the remark this answers is not in the document]
Comment (henryboldi) [reply ui/components/Shared/SharedSlideUpMenu.tsx:45]: That solution sounds good to me, but using a ref for the timeout. The `Snackbar` does something similar:

```js
  const msTillDismiss = 2500

  const snackbarTimeout = useRef<ReturnType<typeof setTimeout>>()

  const clearSnackbarTimeout = () => {
    if (snackbarTimeout.current) {
      clearTimeout(snackbarTimeout.current)
    }
  }

  useEffect(() => {
    clearSnackbarTimeout()

    snackbarTimeout.current = setTimeout(() => {
      dispatch(clearSnackbarMessage())
    }, msTillDismiss)
  }, [snackbarTimeout, snackbarMessage, msTillDismiss, dispatch])

  useEffect(() => {
    window.onblur = () => {
      clearSnackbarTimeout()
      dispatch(clearSnackbarMessage())
    }
  }, [dispatch])
 ```

**Pull request merged**